client/data/media/use-delete-media-mutation.js

```diff
@@ -0,0 +1,24 @@
+import { useCallback } from 'react';
+import { useMutation } from 'react-query';
+import wp from 'calypso/lib/wp';
+
+export const useDeleteMediaMutation = ( siteId, queryOptions = {} ) => {
+	const mutation = useMutation(
+		( { mediaId } ) => wp.req.post( `/sites/${ siteId }/media/${ mediaId }/delete` ),
+		{ ...queryOptions }
+	);
+
+	const { mutate } = mutation;
+
+	const deleteMedia = useCallback(
+		( items ) => {
+			const mediaItems = Array.isArray( items ) ? items : [ items ];
+			mediaItems
+				.filter( ( item ) => typeof item.ID === 'number' )
+				.forEach( ( item ) => mutate( { mediaId: item.ID } ) );
```

Comment:
The react-query docs for the `mutate` function say:

> If you make multiple requests, `onSuccess` will fire only after the latest call you've made.

So, will `onMutate`, `onSuccess` and `onError` really be called for each `mutate` call, with different `mediaId` and `deletedMediaItem`? Or are they somehow squashed together? Worth verifying.

Reply:
I just spent some time looking at the `react-query` source, and it seems that when you call `mutate` multiple times, the hook will stop observing the old query and will start observing the new one. And the `on*` callbacks are triggered by the current (latest) observer. That means only the last `mutate` call will trigger them.

If you want to process all `mutate` results, you'll need to call `mutateAsync` and process the returned promises.

Reply:
For what I can see we have `onSuccess` calls for every deleted item (see screenshot of Redux actions dispatched after deleting 2 items).

<img width="319" alt="Screenshot 2021-12-01 at 10 23 38" src="https://user-images.githubusercontent.com/9202899/144207374-20190fad-3df4-4343-bd16-9628bb147acf.png">

Reply:
I just tried this:
```js
function Mutate() {
  const { mutate } = useMutation(
    async ({ mediaId }) => {
      console.log("starting mutation", mediaId);
      await new Promise((r) => setTimeout(r, 1000));
      console.log("finished mutation", mediaId);
      return mediaId;
    },
    {
      onMutate: ({ mediaId }) => {
        console.log("event starting mutation", mediaId);
      },
      onSuccess: (item, { mediaId }) => {
        console.log("event finished mutation", item, mediaId);
      },
    }
  );

  React.useEffect(() => {
    mutate({ mediaId: 1 });
    mutate({ mediaId: 2 });
    mutate({ mediaId: 3 });
    mutate({ mediaId: 4 });
  }, []);

  return null;
}
```
And it calls `onMutate` and `onSuccess` for all four mutations. So my concerns were not valid. The value of `mediaId` doesn't matter, and the presence of variables doesn't matter either.

Reply:
@jsnajdr I see now where the confusion may have come from. The docs you're referring to regarding the `onSuccess` call is related to the handler you can pass in to the `mutate` function. However, in the code above you're passing `onSuccess` to `useMutation` which seems to be different in that regard. Haven't tested though.

Reply:
You're right, it depends whether `onSuccess` is a `useMutation` option or a `mutate` option. In this example:
```js
function Mutate() {
  const { mutate } = useMutation(async ({ mediaId }) => {
    console.log("starting mutation", mediaId);
    await new Promise((r) => setTimeout(r, 1000));
    console.log("finished mutation", mediaId);
    return mediaId;
  });

  React.useEffect(() => {
    for (let i = 0; i < 4; i++) {
      mutate(
        { mediaId: i },
        {
          onSuccess: (item, { mediaId }) => {
            console.log("event finished mutation", item, mediaId);
          },
        }
      );
    }
  }, []);
  return null;
}
```
`onSuccess` is called only once, for the last mutation:
```
event finished mutation 3
```
I don't know, the behavior looks unreasonable to me. When I specify a _global_ `onSuccess` callback on `useMutation`, it's called every time, but a _local_ callback on `mutate` is called only on the last mutation? That doesn't make sense, local callbacks should influence each other _less_ than the global ones, but here it is the other way around. Let's file a bug 🙂

Reply:
Filed a complaint here 🙂 https://github.com/tannerlinsley/react-query/issues/3040